### PR TITLE
Clean-up Python #4815 - P1.10: Chem\MolDb

### PR DESCRIPTION
**Pull request by IchiruTake**
This is the sub-PR breaking from the large PR 4815: https://github.com/rdkit/rdkit/pull/4815

--------------------------------
Updated Changes:
- [New]: Clean tons of unused import in `Loader_sa.py`

### CODE REVIEW

File: rdkit/Chem/MolDb/Loader_orig.py [hunk 1]
```diff
@@ -70,27 +70,33 @@ def ProcessMol(mol, typeConversions, globalProps, nDone, nameProp='_Name', nameC
...
-    newRow.append(row[2])
+    for counter, pn in enumerate(globalProps): # pn is the key
+      newRow[2 + counter] = pD.get(pn, defaultVal)
+
     if not skipSmiles:
-      newRow.append(row[3])
+      newRow[-2] = row[2]
+      newRow[-1] = row[3]
+    else:
+      newRow[-1] = row[2]
```
Comment (gedeck) [inline rdkit/Chem/MolDb/Loader_orig.py:98]: If we keep this rewrite, reverse the if statement to have a simpler condition (no double negative - not not adding):
```
    if skipSmiles:
      newRow[-1] = row[2]
    else:
      newRow[-2] = row[2]
      newRow[-1] = row[3]
```

**Review by gedeck — CHANGES_REQUESTED** [review]

File: rdkit/Chem/MolDb/Loader_orig.py [hunk 2]
```diff
@@ -52,16 +53,15 @@ def ProcessMol(mol, typeConversions, globalProps, nDone, nameProp='_Name', nameC
...
+    for pn in pns:
       if pn.lower() == nameCol.lower():
         continue
       pv = mol.GetProp(pn).strip()
       if pv.find('>') < 0 and pv.find('<') < 0:
         colTyp = globalProps.get(pn, 2)
         while colTyp > 0:
           try:
-            tpi = typeConversions[colTyp][1](pv)
+            typeConversions[colTyp][1](pv)
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/MolDb/Loader_orig.py:64]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/MolDb/Loader_orig.py [hunk 3]
```diff
@@ -70,27 +70,33 @@ def ProcessMol(mol, typeConversions, globalProps, nDone, nameProp='_Name', nameC
...
-  else:
-    pD = {}
+
   if redraw:
     AllChem.Compute2DCoords(m)
   if not skipSmiles:
     row.append(Chem.MolToSmiles(mol, True))
-  row.append(DbModule.binaryHolder(mol.ToBinary()))
-  row.append(pD)
+  row.extend([DbModule.binaryHolder(mol.ToBinary()), pD])
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/MolDb/Loader_orig.py:78]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/MolDb/Loader_orig.py [hunk 4]
```diff
@@ -70,27 +70,33 @@ def ProcessMol(mol, typeConversions, globalProps, nDone, nameProp='_Name', nameC
...
+    newRow = [0] * (PropsSize + 3 + int(not skipSmiles))
+    newRow[0] = row[0]
+    newRow[1] = row[1]
+    
     pD = row[-1]
-    for pn in globalProps:
-      pv = pD.get(pn, defaultVal)
-      newRow.append(pv)
-    newRow.append(row[2])
+    for counter, pn in enumerate(globalProps): # pn is the key
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/MolDb/Loader_orig.py:91]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/MolDb/Loader_orig.py [hunk 5]
```diff
@@ -70,27 +70,33 @@ def ProcessMol(mol, typeConversions, globalProps, nDone, nameProp='_Name', nameC
...
+      newRow[2 + counter] = pD.get(pn, defaultVal)
+
     if not skipSmiles:
-      newRow.append(row[3])
+      newRow[-2] = row[2]
+      newRow[-1] = row[3]
+    else:
+      newRow[-1] = row[2]
+
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/MolDb/Loader_orig.py:99]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/MolDb/Loader_orig.py (hunk 1 above)
Comment (IchiruTake) [reply rdkit/Chem/MolDb/Loader_orig.py]: OK, I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/Chem/MolDb/Loader_orig.py [hunk 6]
```diff
@@ -158,7 +158,7 @@ def LoadDb(suppl, dbName, nameProp='_Name', nameCol='compound_id', silent=False,
       pass
     curs.execute(f'create table {regName} ({",".join(typs)})')
   else:
-    curs.execute(f'select * from {(regName, )} limit 1')
+    curs.execute(f'select * from {regName} limit 1')
```
Comment (gedeck) [inline rdkit/Chem/MolDb/Loader_orig.py:161]: This change is not identical:
```
>>> regName = 'abc'
>>> f'select * from {(regName, )} limit 1'
"select * from ('abc',) limit 1"
>>> f'select * from {regName} limit 1'
'select * from abc limit 1'
```

File: rdkit/Chem/MolDb/Loader_orig.py [hunk 7]
```diff
@@ -153,12 +153,12 @@ def LoadDb(suppl, dbName, nameProp='_Name', nameCol='compound_id', silent=False,
   curs = conn.GetCursor()
   if startAnew:
     try:
-      curs.execute(f'drop table {regName}' % regName)
+      curs.execute(f'drop table {regName}')
     except Exception:
       pass
-    curs.execute(f'create table {regName} {",".join(typs)}')
+    curs.execute(f'create table {regName} ({",".join(typs)})')
```
Comment (gedeck) [inline rdkit/Chem/MolDb/Loader_orig.py:159]: Not identical:
```
>>> f'create table {regName} {",".join(typs)}'
'create table abc d,e,f'
>>> f'create table {regName} ({",".join(typs)})'
'create table abc (d,e,f)'
```

File: rdkit/Chem/MolDb/Loader_orig.py [hunk 8]
```diff
@@ -153,12 +153,12 @@ def LoadDb(suppl, dbName, nameProp='_Name', nameCol='compound_id', silent=False,
...
   else:
-    curs.execute(f'select * from {regName} limit 1')
+    curs.execute(f'select * from {(regName, )} limit 1')
```
Comment (gedeck) [inline rdkit/Chem/MolDb/Loader_orig.py]: Not identical:
```
>>> f'select * from {regName} limit 1'
'select * from abc limit 1'
>>> f'select * from {(regName, )} limit 1'
"select * from ('abc',) limit 1"
```

File: rdkit/Chem/MolDb/Loader_orig.py [hunk 9]
```diff
@@ -136,49 +129,49 @@ def LoadDb(suppl, dbName, nameProp='_Name', nameCol='compound_id', silent=False,
...
   conn = DbConnect(dbName)
   curs = conn.GetCursor()
   if startAnew:
     try:
-      curs.execute('drop table %s' % regName)
+      curs.execute(f'drop table {regName}' % regName)
     except Exception:
       pass
-    curs.execute('create table %s (%s)' % (regName, ','.join(typs)))
+    curs.execute(f'create table {regName} {",".join(typs)}')
```
Comment (gedeck) [inline rdkit/Chem/MolDb/Loader_orig.py]: Change is not identical:
```
>>> 'create table %s (%s)' % (regName, ','.join(typs))
'create table abc (d,e,f)'
>>> f'create table {regName} {",".join(typs)}'
'create table abc d,e,f'
```

**Review by gedeck — CHANGES_REQUESTED** [review]

**Comment by gedeck** [comment]
@IchiruTake I was looking through the commits and only two of the comments show up in the overall changes, so it could be that you already fixed some of the f-strings.

File: rdkit/Chem/MolDb/Loader_orig.py (hunk 6 above)
Comment (IchiruTake) [reply rdkit/Chem/MolDb/Loader_orig.py:161]: I have tested on GDB online

```python
import sys
import os

print(sys.version, sys.version_info)

regName = 'abc'

x = 'select * from %s limit 1' % (regName, )
print(x)

y = f'select * from {regName} limit 1'
print(y)
```

The result is `x==y` is evaluate to be True
Tested on GDB online and `online-python.com`

You can see the previous commit. It raised error in unit test, and I have to fix twice of it.

File: rdkit/Chem/MolDb/Loader_orig.py (hunk 7 above)
Comment (IchiruTake) [reply rdkit/Chem/MolDb/Loader_orig.py:159]: I don't understand, the original code in line 160 has round bracket. This type of error raised once in the first commit after initial review. You can see below

**Comment by IchiruTake** [comment]
> @IchiruTake I was looking through the commits and only two of the comments show up in the overall changes, so it could be that you already fixed some of the f-strings.

I have responded to two of the latest review, it raised error and I have to make two attempts to dig them out. You can check with that.

File: rdkit/Chem/MolDb/Loader_sa.py [hunk 10]
```diff
@@ -68,7 +59,7 @@ def ProcessMol(session, mol, globalProps, nDone, nameProp='_Name', nameCol='comp
   except KeyError:
     nm = None
   if not nm:
-    nm = 'Mol_%d' % nDone
+    nm = f'Mol_{nDone}' % nDone
```
Comment (gedeck) [inline rdkit/Chem/MolDb/Loader_sa.py:62]: The `% nDone` is redundant.

`nm = f'Mol_{nDone}'`

**Review by gedeck — APPROVED** [review]
@IchiruTake one comment is for a minor issue that needs fixing.

@greglandrum otherwise ready for final review

File: rdkit/Chem/MolDb/Loader_sa.py [hunk 11]
```diff
@@ -34,24 +27,22 @@ class Compound(decBase):
...
-  return maker
+  return sessionmaker(bind=engine)
 
 
 ConnectToSchema = RegisterSchema
 
 
 def _ConnectToSchema(dbUrl, echo=False):
   engine = create_engine(dbUrl, echo=echo)
-  meta
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/MolDb/Loader_sa.py:46]: I have tested when removing this line, and it is fine to remove. Shall I remove it?

File: rdkit/Chem/MolDb/FingerprintUtils.py [hunk 12]
```diff
@@ -52,7 +51,7 @@ def GetQueryText(mol, query=True):
             if not word:
                 continue
             idx = idx + 1
-            colqs.append('%(word)d&Col_%(idx)d=%(word)d' % locals())
+            colqs.append(f'{word}&Col_{idx}={word}')
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/MolDb/FingerprintUtils.py:54]: I have test here and it is fine to refactoring: 

``` python

word, idx = 1, 2
string = '%(word)d&Col_%(idx)d=%(word)d' % locals()
print(string) # '1&Col_2=11

string = f'{word}&Col_{idx}={word}'
print(string) # '1&Col_2=11
```

File: rdkit/Chem/MolDb/Loader_sa.py (hunk 10 above)
Comment (IchiruTake) [reply rdkit/Chem/MolDb/Loader_sa.py:62]: I will fix it in the next patch

File: rdkit/Chem/MolDb/Loader_sa.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (gedeck) [reply rdkit/Chem/MolDb/Loader_sa.py:46]: Let's Greg comment on this. I assume that at some point the method was called `ConnectToSchema` and it had a purpose. Later SQLalchemy changed which made it no longer required (I vaguely remember that I had to  and the code in `RegisterSchema` was sufficient to achieve the same objective. So `ConnectToSchema` was renamed to `_ConnectToSchema` and for compatibility the line `ConnectToSchema = RegisterSchema` was added.

File: rdkit/Chem/MolDb/FingerprintUtils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/Chem/MolDb/FingerprintUtils.py:54]: @gedeck This is the least weird change of f-string manipulation I have made, but this type is relatively easy for refactoring. There are even weirder case I have mentioned in the latest cleanup (#4966).

File: rdkit/Chem/MolDb/Loader_sa.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (greglandrum) [reply rdkit/Chem/MolDb/Loader_sa.py:46]: This code is all so ancient that I really don't remember the origin of this stuff.
Given that the name starts with "_" and that it's not being used anywhere else I think it makes sense to just delete it.

File: rdkit/Chem/MolDb/FingerprintUtils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (greglandrum) [reply rdkit/Chem/MolDb/FingerprintUtils.py:54]: this looks fine to me

File: rdkit/Chem/MolDb/FingerprintUtils.py [hunk 13]
```diff
@@ -86,18 +85,17 @@ def BuildTorsionsFP(mol):
...
+    return Chem.RDKFingerprint(mol, nBitsPerHash=1)
 
 
 def BuildPharm2DFP(mol):
     global sigFactory
     from rdkit.Chem.Pharm2D import Generate
     try:
         fp = Generate.Gen2DFingerprint(mol, sigFactory)
-    except IndexError:
+    except IndexError as e:
```
Comment (greglandrum) [inline rdkit/Chem/MolDb/FingerprintUtils.py:96]: I don't understand this change. Why introduce the `as` if you aren't going to use the exception?
Note that the original code matches the pattern shown in the python docs (second code example here: https://docs.python.org/3/tutorial/errors.html#raising-exceptions)

File: rdkit/Chem/MolDb/Loader_orig.py [hunk 14]
```diff
@@ -70,27 +70,26 @@ def ProcessMol(mol, typeConversions, globalProps, nDone, nameProp='_Name', nameC
...
-    row.append(Chem.MolToSmiles(mol, True))
+    row.append(Chem.MolToSmiles(mol))
   row.append(DbModule.binaryHolder(mol.ToBinary()))
   row.append(pD)
   return row
 
 
 def ConvertRows(rows, globalProps, defaultVal, skipSmiles):
+  #  globalProps is a python dictionary
+  PropsSize: int = len(globalProps)
```
Comment (greglandrum) [inline rdkit/Chem/MolDb/Loader_orig.py:85]: we don't use type hints in RDKit code, please remove this.

File: rdkit/Chem/MolDb/Loader_sa.py (hunk 10 above)
Comment (greglandrum) [reply rdkit/Chem/MolDb/Loader_sa.py:62]: please do

**Review by greglandrum — CHANGES_REQUESTED** [review]
suggested a few changes here

File: rdkit/Chem/MolDb/FingerprintUtils.py (hunk 13 above)
Comment (IchiruTake) [reply rdkit/Chem/MolDb/FingerprintUtils.py]: Thank you for your review. I will revert back into original

File: rdkit/Chem/MolDb/Loader_orig.py (hunk 14 above)
Comment (IchiruTake) [reply rdkit/Chem/MolDb/Loader_orig.py]: Thank you for your review. I will revert back into original

**Review by greglandrum — APPROVED** [review]
LGTM

**Pull request merged**